langchain/src/callbacks/handlers/tracer_langchain.ts

```diff
@@ -0,0 +1,207 @@
+    return this._handleSessionResponse(endpoint);
+  }
+
+  protected async persistSession(
+    sessionCreate: Optional<BaseTracerSessionV2, "tenant_id">
+  ): Promise<TracerSession> {
+    const endpoint = `${this.endpoint}/sessions`;
+    const tenant_id = this.tenantId ?? (await this.updateTenantId());
+    const response = await fetch(endpoint, {
+      method: "POST",
+      headers: this.headers,
+      body: JSON.stringify(sessionCreate),
+    });
+    if (!response.ok) {
+      if (sessionCreate.name !== undefined) {
+        return await this.loadSession(sessionCreate.name);
+      } else {
+        return await this.loadDefaultSession();
+      }
+    }
+    return {
+      id: (await response.json()).id,
+      tenant_id,
+      ...sessionCreate,
+    };
+  }
+
+  async updateTenantId(): Promise<string> {
+    const endpoint = `${this.endpoint}/tenants`;
+    const response = await fetch(endpoint, {
+      method: "GET",
+      headers: this.headers,
+    });
+
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch tenant ID: ${response.status} ${response.statusText}`
+      );
+    }
+
+    const tenants = await response.json();
+    if (!tenants || tenants.length === 0) {
+      throw new Error(`No tenants found for endpoint ${endpoint}`);
+    }
+
+    const tenantId = tenants[0].id;
+    this.tenantId = tenantId;
+    return tenantId;
+  }
+
+  protected async _handleSessionResponse(
+    endpoint: string
+  ): Promise<TracerSession> {
+    const tenantId = this.tenantId ?? (await this.updateTenantId());
+    const configured_endpoint = `${endpoint}&tenant_id=${this.tenantId}`;
+    const response = await fetch(configured_endpoint, {
+      method: "GET",
+      headers: this.headers,
+    });
+    let tracerSession: TracerSession;
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch session: ${response.status} ${response.statusText}`
+      );
+    }
+    const resp = (await response.json()) as TracerSession[];
+    if (resp.length === 0) {
+      tracerSession = {
+        id: uuid.v4(),
+        start_time: Date.now(),
+        tenant_id: tenantId,
+      };
+      this.session = tracerSession;
+      return tracerSession;
+    }
+    [tracerSession] = resp;
+    this.session = tracerSession;
+    return tracerSession;
+  }
+
+  private async _convertToCreate(
+    run: Run,
+    example_id: string | undefined = undefined
+  ): Promise<RunCreate> {
+    const session = this.session ?? (await this.loadDefaultSession());
+    const persistedRun: RunCreate = {
+      id: run.id,
+      name: run.name,
+      start_time: run.start_time,
+      end_time: run.end_time,
+      run_type: run.run_type,
+      reference_example_id: example_id,
+      extra: run.extra ?? {},
+      execution_order: run.execution_order,
+      serialized: run.serialized,
+      error: run.error,
+      inputs: run.inputs,
+      outputs: run.outputs ?? {},
```

Comment:
This is a little strange as runs with error shouldn't have outputs

Reply:
It's supposed to be optional - I can double check that it works, may have added unnecessarily